### DISCUSSION TITLE
Add loggers to schema

### DISCUSSION THREAD

**Discussion by saskiad** [AllenNeuralDynamics/aind-data-schema#978] [Ideas]
Loggers that describe data format headers. Bruno and Josh to provide examples

**Comment by jtyoung84** [comment]
Probably should be an issue in aind-metadata-mapper

**Comment by bruno-f-cruz** [comment]
This might require a longer discussion to try to solve this issue in a more holistic way. Regardless, I will try to start this discussion with some notes:

## The case for flat binary files
People tend to gravitate towards the same formats in neuroscience. For instance, ephys and ophys data is often saved as [flat binary files](https://open-ephys.atlassian.net/wiki/spaces/OEW/pages/166789121/Flat+binary+format), where data can be readily decoded by having access to some logging "metadata". For this specific case, a user would need to know:

Layout: [ColumnMajor, RowMajor]
StructFormat: U16,U16,F32,F64....
If one assumes the file is homogenous (i.e. all values in a row have the same value), the metadata could even be:
Layout: [ColumnMajor, RowMajor]
DataFormat: U16
NumberOfChannels: 16

## CSVs

CSVs are another great example of how people tend to log their data. They have a [defined standard](https://en.wikipedia.org/wiki/Comma-separated_values), and although far from efficient, do offer some human-readability to the user.
A metadata file for such a file could simply contain:
Delimiter: ','
Header: [ColA, ColB, ColC]

## Harp binary files

The harp protocol is a well-defined [binary protocol](https://github.com/harp-tech/protocol/blob/main/BinaryProtocol-8bit.md), with [recommend logging practices](https://harp-tech.org/docs/articles/logging.html). Similary to the binary file, a few metadata fields could suffice:
HarpBoard: [Behavior, AnalogInput, Lickometer, Olfactometer...]
Suffix: The suffix append to the file generated from each register

## Json Files

This one is a bit more obvious, but saving a deserialized object that follows a pre-existing schema is super powerful, especially if one allows arrays of objects in a single file

SchemaSource: url to schema
IsMultiObject: bool

--- 

These are just some examples. Obviously, each of these logger objects should inherit from a base object that should have some core properties:

RootFolder
StreamName
FileName
FileExtension
